lib/agent.js

```diff
@@ -258,12 +265,10 @@ Agent.prototype._open = function() {
       agent._handleMessage(request.data, callback);
     });
   });
-
-  this.stream.on('end', function() {
-    agent.backend.agentsCount--;
-    if (!agent.stream.isServer) agent.backend.remoteAgentsCount--;
-    agent._cleanup();
-  });
+  
+  var cleanup = agent._cleanup.bind(agent);
+  this.stream.on('end', cleanup);
```

Comment:
Suggestion by Nate for the future, paraphrased by me:

> In the future, we should consider _not_ reacting to `'end'`, and only reacting to `'close'`. He says that Share originally reacted to `'end'` since the original stream implementation by Avital emitted both `'end'` and `'close'`, since that was the recommendation for old versions of Node.
> 
> Nowadays, we probably no longer need to do that.